test/device_authorization_grant_test.dart

```diff
@@ -0,0 +1,141 @@
+        Uri.parse('https://example.com/token'),
+        secret: 'secret',
+        basicAuth: false,
+        httpClient: client);
+  });
+
+  group('.getDeviceCode', () {
+    test('sends an code request', () async {
+      client.expectRequest((request) {
+        expect(request.method, equals('POST'));
+        expect(request.url.toString(), equals(grant.deviceEndpoint.toString()));
+        expect(
+            request.bodyFields,
+            allOf([
+              containsPair('client_id', 'identifier'),
+              containsPair('scope', 'test_scope')
+            ]));
+
+        return Future.value(http.Response(
+            jsonEncode({
+              'device_code': 'deviceCode',
+              'user_code': 'userCode',
+              'verification_uri': 'https://examle.com/verify',
+              'expires_in': 1234,
+            }),
+            200,
+            headers: {'content-type': 'application/json'}));
+      });
+
+      expect(
+          await grant.getDeviceCode(scopes: ['test_scope']),
+          isA<DeviceProperties>()
+              .having((c) => c.device_code, 'device_code', 'deviceCode'));
+    });
+
+    test("can't be called twice", () async {
+      client.expectRequest((request) => Future.value(http.Response(
+          jsonEncode({
+            'device_code': 'deviceCode',
+            'user_code': 'userCode',
+            'verification_uri': 'https://examle.com/verify',
+            'expires_in': 1234,
+          }),
+          200,
+          headers: {'content-type': 'application/json'})));
+      await grant.getDeviceCode();
+      expect(() async => await grant.getDeviceCode(), throwsStateError);
+    });
+  });
+
+  group('.pollForToken', () {
+    test('sends an access token request', () async {
+      client.expectRequest((request) => Future.value(http.Response(
+          jsonEncode({
+            'device_code': 'deviceCode',
+            'user_code': 'userCode',
+            'verification_uri': 'https://examle.com/verify',
+            'expires_in': 1234,
+          }),
+          200,
+          headers: {'content-type': 'application/json'})));
+      await grant.getDeviceCode();
+
+      client.expectRequest((request) {
+        expect(request.method, equals('POST'));
+        expect(request.url.toString(), equals(grant.tokenEndpoint.toString()));
+
+        expect(
+            request.bodyFields,
+            allOf([
+              containsPair('client_id', 'identifier'),
+              containsPair('device_code', 'deviceCode'),
+              containsPair(
+                  'grant_type', 'urn:ietf:params:oauth:grant-type:device_code'),
+            ]));
+
+        return Future.value(http.Response(
+            jsonEncode({
+              'access_token': 'access token',
+              'token_type': 'bearer',
+            }),
+            200,
+            headers: {'content-type': 'application/json'}));
+      });
+
+      expect(
+          await grant.pollForToken(),
```

Comment:
Don't we need a test for what happens if `grant.pollForToken()` doesn't succeed, because the user haven't authenticated the device yet... and the application is supposed to call `grant.pollForToken()` again.